lib/src/merged_tree.rs

```diff
@@ -791,10 +768,64 @@ impl Iterator for TreeDiffIterator<'_> {
     }
 }
 
+impl TreeDiffDirItem {
+    fn from_trees(
+        dir: &RepoPath,
+        trees1: &Merge<Tree>,
+        trees2: &Merge<Tree>,
+        matcher: &dyn Matcher,
+    ) -> Self {
+        let mut entries = vec![];
+        for (name, before, after) in merged_tree_entry_diff(trees1, trees2) {
+            let path = dir.join(name);
+            let before = before.to_merge();
+            let after = after.to_merge();
+            let tree_before = before.is_tree();
+            let tree_after = after.is_tree();
+            // Check if trees and files match, but only if either side is a tree or a file
+            // (don't query the matcher unnecessarily).
+            let tree_matches = (tree_before || tree_after) && !matcher.visit(&path).is_nothing();
+            let file_matches = (!tree_before || !tree_after) && matcher.matches(&path);
+
+            // Replace trees or files that don't match by `Merge::absent()`
+            let before = if (tree_before && tree_matches) || (!tree_before && file_matches) {
+                before
+            } else {
+                Merge::absent()
+            };
+            let after = if (tree_after && tree_matches) || (!tree_after && file_matches) {
+                after
+            } else {
+                Merge::absent()
+            };
+            if before.is_absent() && after.is_absent() {
+                continue;
+            }
+            entries.push(TreeDiffEntry {
+                source: path.clone(),
+                target: path,
+                value: Ok((before, after)),
+            });
+        }
+        entries.reverse();
+        Self { entries }
+    }
+}
+
+impl Iterator for TreeDiffIterator<'_> {
+    type Item = TreeDiffEntry;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next_impl()
+            .map(|e| e.adjust_for_copy_tracking(&self.source, self.copy_records))
+    }
+}
```

Comment:
Could you add some tests in `test_merged_tree.rs`? Please also include tests for things that are not handled well yet and just leave a TODO in the test to say that we need to fix it later.

In case you're wondering what the advantages are compared to CLI tests:

* Can add comprehensive tests and then not need so comprehensive tests for each command that displays diffs (or not have to arbitrarily decide which of the commands should get comprehensive tests)
* Easier to produce corner cases (for example, you can pass in a `CopyRecordMap` you can't trigger via Git
* Faster

Reply:
I have added some tests in `test_merged_tree.rs`.  There aren't really many corner cases that can be triggered at this level.  The current api of `CopyRecords` doesn't allow conflicts and those are the cases that have corners and need to be designed.